Handle `read`s on STDIN

samrat opened the pull request:
Closes #1434

RalfJung commented on src/shims/foreign_items/posix.rs:
```diff
@@ -66,7 +66,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                    use std::io::{self, Read};
+
+                    let mut buffer = String::new();
+                    let res = io::stdin().read_to_string(&mut buffer);
```
This does not look right -- you should somehow be taking `count` into account, or you might be reading more than fits the target buffer.

RalfJung replied:
Your `write_bytes` could cause memory corruption in the interpreted program because you are writing beyond the bounds of the buffer you were given.

Another problem here is that `read_to_string` will fail on non-UTF-8 data. We are working with raw bytes here, so the right type to use is `Vec<u8>`.

RalfJung replied:
You can use the `read` function in `fs.rs` for inspiration.

RalfJung commented on src/shims/foreign_items/posix.rs:
```diff
@@ -66,7 +66,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                            this.memory.write_bytes(buf, buffer.bytes())?;
+                            i64::try_from(bytes).unwrap()
+                        },
+                        Err(_) => -1,
```
When returning "-1", we have to set errno to an appropriate error... but I realize the stdout code also does not do that. So maybe add a FIXME in both places so we don't forget to do this eventually?

RalfJung commented on src/shims/foreign_items/posix.rs:
```diff
@@ -66,7 +66,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 let count = this.read_scalar(count)?.to_machine_usize(this)?;
                 let result = if fd == 0 {
-                    throw_unsup_format!("reading from stdin is not implemented")
+                    use std::io::{self, Read};
```
Since reading from stdin can break non-determinism, please add a call to `this.check_no_isolation`.

RalfJung commented:
@samrat FYI, this PR has some feedback for you to address. :)  No rush though, just making sure it does not get lost.

bors commented:
:umbrella: The latest upstream changes (presumably #1460) made this pull request unmergeable. Please resolve the merge conflicts.

RalfJung commented:
@samrat what is the status here? Do you still plan to work on this?

samrat commented:
@RalfJung Apologies for the lack of updates here, and for the delay. I'll try to work on it this weekend and address the feedback on the PR.

RalfJung commented:
No need to rush, I'd just like to close this PR and give other contributors a chance in case you decide to not work on this for the foreseeable future. :) But

RalfJung commented on src/shims/posix/foreign_items.rs:
```diff
@@ -67,7 +67,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
                 let count = this.read_scalar(count)?.to_machine_usize(this)?;
                 let result = if fd == 0 {
-                    throw_unsup_format!("reading from stdin is not implemented")
+                    use std::io::{self, Read};
```
Please add this at the top of the file, we usually avoid local imports.

RalfJung commented on src/shims/posix/foreign_items.rs:
```diff
@@ -67,7 +67,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+
+                    this.check_no_isolation("read")?;
+
+                    let mut buffer = vec![0; count as usize];
```
`as` casts are always suspicious. Why can this not be bigger than a host usize?
This is what the `read` for files does:
```rust
// We cap the number of read bytes to the largest value that we are able to fit in both the
// host's and target's `isize`. This saves us from having to handle overflows later.
let count = count.min(this.machine_isize_max() as u64).min(isize::MAX as u64);
```

RalfJung commented on src/shims/posix/foreign_items.rs:
```diff
@@ -67,7 +67,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                            this.memory.write_bytes(buf, buffer)?;
+                            i64::try_from(bytes).unwrap()
+                        },
+                        // FIXME: set errno to appropriate value
```
Setting the errno is easy:
```rust
                Err(e) => {
                    this.set_last_error_from_io_error(e)?;
                    Ok(-1)
                }
```

RalfJung commented:
Does anyone reading along have an idea for how to properly test this?

samrat commented:
@RalfJung I've addressed your comments. 

This does introduce some duplication with `read`, so I attempted factoring out the common parts into a method that takes `impl Read` and so can read both `File`'s and `Stdin`: https://github.com/samrat/miri/commit/3b9c8d3cd13bf3ad6a4badbc90ab3bc880d6dba2#diff-560909e2fdd6fac0fb0b832d9470ac7cR438. Currently, I'm not sure how to resolve a multiple mutable borrows type error that results with this approach.

Do you think this approach is workable?

RalfJung commented on src/shims/posix/foreign_items.rs:
```diff
@@ -67,10 +69,23 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                    let count = count.min(this.machine_isize_max() as u64).min(isize::MAX as u64);
+                    // This can never fail because `count` was capped
+                    // to be smaller than `isize::MAX`.
+                    let count = isize::try_from(count).unwrap();
```
This is unnecessary (and I recently removed it from `fs.rs`).

RalfJung commented:
> This does introduce some duplication with read

True. Let's go with this for the PR though, and do refactoring later.
IMO the proper fix here is to handle FDs 1-3 in `fs.rs` together with the rest of them.

RalfJung commented on src/shims/posix/foreign_items.rs:
```diff
@@ -11,6 +12,7 @@ use shims::posix::fs::EvalContextExt as _;
 use shims::posix::sync::EvalContextExt as _;
 use shims::posix::thread::EvalContextExt as _;
 
+
```
Spurious new newline.

RalfJung commented:
Thanks. :)
@bors r+

bors commented:
:pushpin: Commit f4d1841811c260c96c71a41774d3cb61b62dad7f has been approved by `RalfJung`

<!-- @bors r=RalfJung f4d1841811c260c96c71a41774d3cb61b62dad7f -->
<!-- homu: {"type":"Approved","sha":"f4d1841811c260c96c71a41774d3cb61b62dad7f","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit f4d1841811c260c96c71a41774d3cb61b62dad7f with merge b021209b0e889d9c61ec41c4016393f7415e69e8...
<!-- homu: {"type":"BuildStarted","head_sha":"f4d1841811c260c96c71a41774d3cb61b62dad7f","merge_sha":"b021209b0e889d9c61ec41c4016393f7415e69e8"} -->

RalfJung commented:
For the `read` refactoring, if you want to work on that, could you open an issue? I am on vacation now, but I can write up mentoring instructions in about a week or so when I am back.

bors commented:
:sunny: Test successful - [checks-travis](https://travis-ci.com/github/rust-lang/miri/builds/176158632), [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/miri/builds/34172398)
Approved by: RalfJung
Pushing b021209b0e889d9c61ec41c4016393f7415e69e8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"status-appveyor":"https://ci.appveyor.com/project/rust-lang-libs/miri/builds/34172398","checks-travis":"https://travis-ci.com/github/rust-lang/miri/builds/176158632"},"merge_sha":"b021209b0e889d9c61ec41c4016393f7415e69e8"} -->
